File: zul/src/main/resources/web/js/zul/mesh/Auxhead.ts
/* Auxhead.ts

	Purpose:

	Description:

	History:
		Mon May  4 15:57:46     2009, Created by tomyeh

Copyright (C) 2009 Potix Corporation. All Rights Reserved.

This program is distributed under LGPL Version 2.1 in the hope that
it will be useful, but WITHOUT ANY WARRANTY.
*/
/**
 * Used to define a collection of auxiliary headers ({@link Auxheader}).
 *
 * <p>Non XUL element.
 * @defaultValue {@link getZclass}: z-auxhead.
 */
@zk.WrapClass('zul.mesh.Auxhead')
export class Auxhead extends zul.mesh.HeadWidget {
	/** @internal */
	override bind_(desktop: zk.Desktop | undefined, skipper: zk.Skipper | undefined, after: CallableFunction[]): void {
		super.bind_(desktop, skipper, after);
		// B50-3306729: the first header should have border-left when the first column is covered with other header
		this.fixBorder_();
	}

	// B50-3306729: the first header should have border-left when the first column is covered with other header
	/** @internal */
	fixBorder_(): void {
		var fc = jq(this).children(':first-child'),
			rspan = fc.attr('rowspan')!,
			times = parseInt(rspan) - 1;
		if (rspan && times > 0) {
			for (var head = this.nextSibling; head && times != 0; head = head.nextSibling, times--)
				jq(head.firstChild!).addClass(this.$s('border'));
		}
	}
}

File: zul/src/main/java/org/zkoss/zul/Auxhead.java
/* Auxhead.java

	Purpose:
		
	Description:
		
	History:
		Wed Oct 24 09:55:47     2007, Created by tomyeh

Copyright (C) 2007 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under LGPL Version 2.1 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/
package org.zkoss.zul;

import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zul.impl.MeshElement;
import org.zkoss.zul.impl.XulElement;

/**
 * Used to define a collection of auxiliary headers ({@link Auxheader}).
 *
 * <p>Non XUL element.
 * <p>Default {@link #getZclass}: z-auxhead.(since 3.5.0)
 * 
 * @since 3.0.0
 * @author tomyeh
 */
public class Auxhead extends XulElement {
	public Auxhead() {
	}

	public String getZclass() {
		return _zclass == null ? "z-auxhead" : _zclass;
	}

	//super//
	public void beforeChildAdded(Component child, Component refChild) {
		if (!(child instanceof Auxheader))
			throw new UiException("Unsupported child: " + child);
		super.beforeChildAdded(child, refChild);
	}

	public void beforeParentChanged(Component parent) {
		super.beforeParentChanged(parent);
		String cssFlexAutoDisabledCountAttr = MeshElement.CSS_FLEX_AUTO_DISABLED_COUNT;
		//handle new parent
		if (parent != null) {
			Integer cssFlexAutoDisabledCount = (Integer) parent.getAttribute(cssFlexAutoDisabledCountAttr);
			if (cssFlexAutoDisabledCount == null) {
				parent.setAttribute(cssFlexAutoDisabledCountAttr, 1);
				parent.invalidate();
			} else {
				parent.setAttribute(cssFlexAutoDisabledCountAttr, ++cssFlexAutoDisabledCount);
			}
		}
		//handle old parent
		Component oldParent = this.getParent();
		if (oldParent != null) {
			Integer oldParentCssFlexAutoDisabledCount = (Integer) oldParent.getAttribute(cssFlexAutoDisabledCountAttr);
			if (oldParentCssFlexAutoDisabledCount == 1) {
				oldParent.removeAttribute(cssFlexAutoDisabledCountAttr);
				oldParent.invalidate();
			} else {
				oldParent.setAttribute(cssFlexAutoDisabledCountAttr, --oldParentCssFlexAutoDisabledCount);
			}
		}
	}
}
